### PR TITLE
feat(hyd): connecting hydraulics to electrical system

### PR DESCRIPTION
**Pull request by crocket63**
## Summary of Changes

This PR will start connecting electrical power to the various hydraulic elements, like electric pumps or various solenoid valves.

- [ ] Engineers successfully connected yellow electric pump power
- [ ] Engineers successfully connected blue electric pump power
- [ ] Engineers successfully connected yellow electric pump control signal
- [ ] Engineers successfully connected blue electric pump control signal
- [ ] Engineers successfully connected engine driven pumps solenoid valves
- [ ] Engineers successfully connected PTU solenoid valves
- [ ] Engineers successfully connected RAT deployment solenoids

## Additional context

Discord username (if different from GitHub):

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: src/systems/a320_systems/src/hydraulic.rs [hunk 1]
```diff
@@ -10,12 +10,16 @@ use systems::hydraulic::{
...
+    AutoOffFaultPushButton, AutoOnFaultPushButton, FirePushButton, MomentaryPushButton,
+    OnOffFaultPushButton,
 };
 use systems::simulation::{
     SimulationElement, SimulationElementVisitor, SimulatorReader, SimulatorWriter, UpdateContext,
 };
 
+use super::A320EmergencyElectricalOverheadPanel;
+
+use systems::electrical::{consumption::SuppliedPower, ElectricalBusType};
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:22]: **suggestion**: Although I'm aware that some of these electrical types currently _must_ be imported. We should try to reduce inter-module dependencies, where e.g. HYD knows about a specific type in ELEC and vice versa. For this reason, I'd suggest adding a trait containing the `rat_and_emer_gen_man_on` fn, and then passing the trait instead of the concrete overhead panel instance.

Overall I try to silo things as much as possible: thus A320 ELEC can use other A320 ELEC types and `systems::electrical` types. A320 HYD can use other A320 HYD types and `systems::hydraulic`. I might move SuppliedPower and ElectricalBusType somewhere else eventually 😅.

File: src/systems/a320_systems/src/hydraulic.rs [hunk 2]
```diff
@@ -189,7 +199,12 @@ impl A320Hydraulic {
             time_to_catch.as_secs_f64() / min_hyd_loop_timestep.as_secs_f64();
 
         // Here we update everything requiring same refresh as the sim calls us, more likely visual stuff
-        self.update_at_every_frames(&context);
+        self.update_at_every_frames(
+            &context,
+            &overhead_panel,
+            &emergency_electrical_overhead,
+            is_in_emergency_elec,
+        );
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:207]:
```suggestion
        self.update_every_frame(
            &context,
            &overhead_panel,
            &emergency_electrical_overhead,
            is_in_emergency_elec,
        );
```

File: src/systems/a320_systems/src/hydraulic.rs [hunk 3]
```diff
@@ -537,9 +566,25 @@ struct A320EngineDrivenPumpController {
...
+        let power_buses;
+        if engine_number == 1 {
+            power_buses = vec![Self::GREEN_EDP_POWER_BUS];
+        } else if engine_number == 2 {
+            power_buses = vec![Self::YELLOW_EDP_POWER_BUS1, Self::YELLOW_EDP_POWER_BUS2];
+        } else {
+            panic!("The A320 only supports two engines.");
+        }
+
+        assert!(!power_buses.is_empty() && power_buses.len() <= 2);
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:583]: **suggestion**: Pass these in via the constructor to save yourself this `if else if else` hassle.
```rust
A320EngineDrivenPumpController::new(1, vec![ElectricalBusType::DirectCurrentEssential]);
A320EngineDrivenPumpController::new(2, vec![ElectricalBusType::DirectCurrent(2), ElectricalBusType::DirectCurrentEssential]);
```

File: src/systems/a320_systems/src/hydraulic.rs [hunk 4]
```diff
@@ -527,6 +554,8 @@ impl HydraulicLoopController for A320HydraulicLoopController {
 }
 
 struct A320EngineDrivenPumpController {
+    is_powered: bool,
+    powering_buses: Vec<ElectricalBusType>,
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:558]: The controller is _powered by_ an electrical bus.

```suggestion
    powered_by: Vec<ElectricalBusType>,
```

File: src/systems/a320_systems/src/hydraulic.rs [hunk 5]
```diff
@@ -617,19 +666,41 @@ impl SimulationElement for A320EngineDrivenPumpController {
...
+        } else if self.engine_number == 2 {
+            self.is_powered = supplied_power
+                .potential_of(&self.powering_buses[0])
+                .is_powered()
+                || supplied_power
+                    .potential_of(&self.powering_buses[1])
+                    .is_powered();
+        } else {
+            panic!("The A320 only supports two engines.");
+        }
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:684]: Just hit the button:

```suggestion
        self.is_powered = self
            .powering_buses
            .iter()
            .any(|bus| supplied_power.potential_of(&bus).is_powered());
```

File: src/systems/a320_systems/src/hydraulic.rs [hunk 6]
```diff
@@ -617,19 +666,41 @@ impl SimulationElement for A320EngineDrivenPumpController {
...
+    }
 }
 
 struct A320BlueElectricPumpController {
+    is_powered: bool,
+    powering_bus: ElectricalBusType,
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:690]: As above.

File: src/systems/a320_systems/src/hydraulic.rs [hunk 7]
```diff
@@ -617,19 +666,41 @@ impl SimulationElement for A320EngineDrivenPumpController {
...
     should_pressurise: bool,
     has_pressure_low_fault: bool,
     is_pressure_low: bool,
     weight_on_wheels: bool,
 }
 impl A320BlueElectricPumpController {
     const MIN_ENGINE_OIL_PRESS_THRESHOLD_TO_INHIBIT_FAULT: f64 = 18.;
+    const POWER_BUS: ElectricalBusType = ElectricalBusType::DirectCurrentEssential;
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:698]: For consistency I'd still suggest injecting this into the type. What do you think?

In comments below I've written "As above".

File: src/systems/a320_systems/src/hydraulic.rs [hunk 8]
```diff
@@ -716,6 +794,8 @@ impl Default for A320BlueElectricPumpController {
 }
 
 struct A320YellowElectricPumpController {
+    is_powered: bool,
+    powering_bus: ElectricalBusType,
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:798]: As above.

File: src/systems/a320_systems/src/hydraulic.rs [hunk 9]
```diff
@@ -725,8 +805,12 @@ impl A320YellowElectricPumpController {
     const DURATION_OF_YELLOW_PUMP_ACTIVATION_AFTER_CARGO_DOOR_OPERATION: Duration =
         Duration::from_secs(20);
 
+    const POWER_BUS: ElectricalBusType = ElectricalBusType::DirectCurrent(2);
+
     fn new() -> Self {
         Self {
+            is_powered: false,
+            powering_bus: Self::POWER_BUS,
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:813]: As above.

File: src/systems/a320_systems/src/hydraulic.rs [hunk 10]
```diff
@@ -791,6 +880,8 @@ impl Default for A320YellowElectricPumpController {
 }
 
 struct A320PowerTransferUnitController {
+    is_powered: bool,
+    powering_bus: ElectricalBusType,
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:884]: As above.

File: src/systems/a320_systems/src/hydraulic.rs [hunk 11]
```diff
@@ -805,8 +896,12 @@ impl A320PowerTransferUnitController {
     const DURATION_AFTER_WHICH_NWS_PIN_IS_REMOVED_AFTER_PUSHBACK: Duration =
         Duration::from_secs(15);
 
+    const POWER_BUS: ElectricalBusType = ElectricalBusType::DirectCurrentGndFltService;
+
     fn new() -> Self {
         Self {
+            is_powered: false,
+            powering_bus: Self::POWER_BUS,
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:904]: As above.

File: src/systems/a320_systems/src/hydraulic.rs [hunk 12]
```diff
@@ -840,13 +935,16 @@ impl A320PowerTransferUnitController {
...
-        self.should_enable = overhead_panel.ptu_push_button_is_auto()
+        let should_enable_if_powered = overhead_panel.ptu_push_button_is_auto()
             && (!self.weight_on_wheels
                 || self.eng_1_master_on && self.eng_2_master_on
                 || !self.eng_1_master_on && !self.eng_2_master_on
                 || (!self.parking_brake_lever_pos
                     && !self.nose_wheel_steering_pin_inserted.output()))
             && !ptu_inhibited;
+
+        // This is negative logic here. If no power, ptu is laways ON
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:946]:
```suggestion
        // When there is no power, the PTU is always ON.
```

File: src/systems/a320_systems/src/hydraulic.rs [hunk 13]
```diff
@@ -866,29 +964,54 @@ impl SimulationElement for A320PowerTransferUnitController {
...
+    fn update(
+        &mut self,
+        overhead_panel: &A320HydraulicOverheadPanel,
+        emergency_electrical_overhead: &A320EmergencyElectricalOverheadPanel,
+        is_in_emergency_elec: bool,
+    ) {
+        let should_solenoid_1_deploy_if_powered = overhead_panel.rat_man_on_push_button_is_on();
+
+        let should_solenoid_2_deploy_if_powered =
+            is_in_emergency_elec || emergency_electrical_overhead.rat_and_emer_gen_man_on();
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:1011]: **suggestion**: Rename to `solenoid_1_should_trigger_deploy(ment)`; as the current name suggests the solenoid is deployed, not the RAT.

Comment (davidwalschots) [reply src/systems/a320_systems/src/hydraulic.rs:1011]: **thought**: Maybe the solenoids should have their own type, so they can handle their output and being powered internally?

File: src/systems/a320_systems/src/hydraulic.rs [hunk 14]
```diff
@@ -1293,7 +1430,19 @@ mod tests {
             hydraulics: A320Hydraulic,
             overhead: A320HydraulicOverheadPanel,
             engine_fire_overhead: A320EngineFireOverheadPanel,
+            emergency_electrical_overhead: A320EmergencyElectricalOverheadPanel,
             landing_gear: LandingGear,
+
+            // Electric buses states to be able to kill them dynamically
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:1436]: **thought**: I think the field names are clear enough (though I would suggest changing `alive` to `powered` to be in line with all the other naming) and don't require a comment.

File: src/systems/a320_systems/src/hydraulic.rs [hunk 15]
```diff
@@ -1349,21 +1512,158 @@ mod tests {
...
+            fn ac_ground_service_power_state(&mut self, bus_is_alive: bool) {
+                self.is_ac_ground_service_alive = bus_is_alive;
+            }
+
+            fn dc_bus_2_power_state(&mut self, bus_is_alive: bool) {
+                self.is_dc_2_alive = bus_is_alive;
+            }
+            fn dc_ess_power_state(&mut self, bus_is_alive: bool) {
+                self.is_dc_ess_alive = bus_is_alive;
+            }
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:1536]: **issue**: Rust setters should start with `set_`.
**suggestion**: Write:
```rust
fn set_dc_bus_2_is_powered(&mut self, is_powered: bool) {
    self.dc_bus_2_is_powered = is_powered;
}
```
**issue**: Inconsistent (missing) whitespace between functions.

File: src/systems/a320_systems/src/hydraulic.rs [hunk 16]
```diff
@@ -2644,6 +3005,73 @@ mod tests {
             assert!(test_bed.yellow_pressure() < Pressure::new::<psi>(500.));
         }
 
+        #[test]
+        fn yellow_edp_solenoid_unpowered() {
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:3009]: **issue**: As in the previous PR, these names indicate what you are testing.
**suggestion**: Use function names that describe what you are testing, e.g.:
`when_yellow_edp_solenoid_unpowered_yellow_hydraulic_system_is_pressurised`

File: src/systems/a320_systems/src/hydraulic.rs [hunk 17]
```diff
@@ -2644,6 +3005,73 @@ mod tests {
...
+                .run_waiting_for(Duration::from_secs(15));
+
+            assert!(test_bed.is_yellow_pressurised());
+
+            // Stoping EDP
+            test_bed = test_bed
+                .set_yellow_ed_pump(false)
+                .run_waiting_for(Duration::from_secs(15));
+
+            assert!(!test_bed.is_yellow_pressurised());
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:3028]: I don't know if these are part of the same test, or if you're actually intending to test multiple scenarios.

File: src/systems/a320_systems/src/hydraulic.rs [hunk 18]
```diff
@@ -2644,6 +3005,73 @@ mod tests {
...
+
+            test_bed = test_bed
+                .dc_bus_2_lost()
+                .run_waiting_for(Duration::from_secs(15));
+
+            // Yellow solenoid has backup power from DC ESS BUS
+            assert!(!test_bed.is_yellow_pressurised());
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:3035]: `when_yellow_edp_turned_off_and_yellow_edp_solenoid_primary_power_source_unpowered_yellow_hydraulic_system_remains_unpressurised`

etc etc

File: src/systems/a320_systems/src/hydraulic.rs [hunk 19]
```diff
@@ -3342,6 +3770,7 @@ mod tests {
             overhead_panel.blue_epump_override_push_button.push_off();
 
             let mut blue_epump_controller = A320BlueElectricPumpController::new();
+            blue_epump_controller.is_powered = true;
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:3773]: **issue**: Breaking encapsulation.
**suggestion**: If you want a pump to be powered, provide power! 😉

File: src/systems/a320_systems/src/hydraulic.rs [hunk 20]
```diff
@@ -3482,6 +3914,19 @@ mod tests {
...
+            blue_epump_controller.is_powered = false;
+            blue_epump_controller.update(
+                &overhead_panel,
+                false,
+                engine_off_oil_pressure,
+                engine_off_oil_pressure,
+                eng1_above_idle,
+                eng2_above_idle,
+            );
+            assert!(!blue_epump_controller.should_pressurise());
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:3929]: Separate test

File: src/systems/a320_systems/src/hydraulic.rs [hunk 21]
```diff
@@ -3505,6 +3951,12 @@ mod tests {
             overhead_panel.yellow_epump_push_button.push_auto();
             yellow_epump_controller.update(&context, &overhead_panel, &fwd_door, &aft_door, true);
             assert!(!yellow_epump_controller.should_pressurise());
+
+            // If unpowered no possible control of the pump
+            yellow_epump_controller.is_powered = false;
+            overhead_panel.yellow_epump_push_button.push_on();
+            yellow_epump_controller.update(&context, &overhead_panel, &fwd_door, &aft_door, true);
+            assert!(!yellow_epump_controller.should_pressurise());
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:3959]: Separate test

File: src/systems/a320_systems/src/hydraulic.rs [hunk 22]
```diff
@@ -3766,17 +4225,107 @@ mod tests {
             assert!(test_bed.get_rat_position() <= 0.);
             assert!(test_bed.get_rat_rpm() <= 1.);
 
-            // Stopping both engines
+            // Killing AC buses
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:4228]: Stating the obvious.

File: src/systems/a320_systems/src/hydraulic.rs [hunk 23]
```diff
@@ -3766,17 +4225,107 @@ mod tests {
...
+            let mut test_bed = test_bed_with()
+                .set_cold_dark_inputs()
+                .in_flight()
+                .start_eng1(Ratio::new::<percent>(80.))
+                .start_eng2(Ratio::new::<percent>(80.))
+                .run_waiting_for(Duration::from_secs(10));
+
+            assert!(!test_bed.get_rat_deploy_command());
+
+            // Killing AC1 bus
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:4250]: Stating the obvious

File: src/systems/a320_systems/src/hydraulic.rs [hunk 24]
```diff
@@ -3766,17 +4225,107 @@ mod tests {
...
+            test_bed = test_bed
+                .ac_bus_1_lost()
+                .run_waiting_for(Duration::from_secs(2));
+
+            // RAT has not deployed
+            assert!(!test_bed.get_rat_deploy_command());
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:4256]: **suggestion** rename to `rat_deploy_commanded` and then remove the comment.

File: src/systems/a320_systems/src/hydraulic.rs [hunk 25]
```diff
@@ -3766,17 +4225,107 @@ mod tests {
...
+
+            //Now all AC off
+            test_bed = test_bed
+                .ac_bus_1_lost()
+                .ac_bus_2_lost()
+                .run_waiting_for(Duration::from_secs(2));
+
+            // RAT avail
+            assert!(test_bed.get_rat_deploy_command());
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:4265]: **suggestion**: As above.

Comment (davidwalschots) [reply src/systems/a320_systems/src/hydraulic.rs:4265]: General note: the `test_bed` structure was set up so that tests describe themselves very well without using comments. Comments are only needed when describing some edge-case for which you have to provide functional information to the dev.

**Review by davidwalschots — CHANGES_REQUESTED** [review]

File: src/systems/a320_systems/src/hydraulic.rs (hunk 1 above)
Comment (crocket63) [reply src/systems/a320_systems/src/hydraulic.rs]: Did something different from your example I saw after doing it. Just tell me what you think.

File: src/systems/a320_systems/src/hydraulic.rs (hunk 5 above)
Comment (crocket63) [reply src/systems/a320_systems/src/hydraulic.rs:684]: Beautiful :)

File: src/systems/a320_systems/src/hydraulic.rs (hunk 7 above)
Comment (crocket63) [reply src/systems/a320_systems/src/hydraulic.rs:698]: Did that for the EDP, while still having the buses as constant in the A320Hydraulics type. Just to have those definition at the same place instead of having various buses names spread around in the constructors calls

File: flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/model/A320_NEO_INTERIOR.xml [hunk 26]
```diff
@@ -1883,13 +1883,14 @@
...
-                        <NODE_ID>LOCK_OVHD_HYD_RATMANON</NODE_ID>
-                        <ANIM_NAME>LOCK_OVHD_HYD_RATMANON</ANIM_NAME>
-                        <WWISE_EVENT_1>apu_generator_switch_on</WWISE_EVENT_1>
-                        <WWISE_EVENT_2>apu_generator_switch_off</WWISE_EVENT_2>
-                        <SWITCH_POSITION_TYPE>L</SWITCH_POSITION_TYPE>
-                        <SWITCH_POSITION_VAR>A32NX_HYD_RATMANONLOCK_TOGGLE</SWITCH_POSITION_VAR>
+                    <UseTemplate Name="FBW_Covered_Push_Toggle">
+                        <NODE_ID>PUSH_OVHD_HYD_RATMANON</NODE_ID>
+                        <LOCK_NODE_ID>LOCK_OVHD_HYD_RATMANON</LOCK_NODE_ID>
+                        <LEFT_SINGLE_CODE>1 (&gt;L:A32NX_OVHD_HYD_RAT_MAN_ON_IS_PRESSED)</LEFT_SINGLE_CODE>
```
[the remark this answers is not in the document]
Comment (crocket63) [reply flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/model/A320_NEO_INTERIOR.xml:1889]: I already noticed the problem but didn't know where to look :D  Thanks for this!  Actually I just pasted this from the elec rat button I think so issue might be the same there.

[the remark this answers is not in the document]
Comment (crocket63) [reply flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/model/A320_NEO_INTERIOR.xml:1889]: But... as I still don't speak polish :D  What is the correct thing to do?
<LEFT_SINGLE_CODE>(&gt;L:A32NX_OVHD_HYD_RAT_MAN_ON_IS_PRESSED)</LEFT_SINGLE_CODE>
This?

[the remark this answers is not in the document]
Comment (Saschl) [reply flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/model/A320_NEO_INTERIOR.xml:1889]: Probably something like `<LEFT_SINGLE_CODE>(L:A32NX_OVHD_HYD_RAT_MAN_ON_IS_PRESSED) ! (>L:A32NX_OVHD_HYD_RAT_MAN_ON_IS_PRESSED)</LEFT_SINGLE_CODE>` ?

[the remark this answers is not in the document]
Comment (davidwalschots) [reply flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/model/A320_NEO_INTERIOR.xml:1889]: It's not so difficult to figure this out:
1. Find the `FBW_Covered_Push_Toggle` template.
2. Note that it uses the `FBW_Push_Toggle` template.
3. Note that `FBW_Push_Toggle` has a `HOLD_SIMVAR` argument which is the name of the simvar to set when the button is held down.

```suggestion
                        <HOLD_SIMVAR>L:A32NX_OVHD_HYD_RAT_MAN_ON_IS_PRESSED</HOLD_SIMVAR>
```

[the remark this answers is not in the document]
Comment (lukecologne) [reply flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/model/A320_NEO_INTERIOR.xml:1889]: FBW_Push_Toggle doesn't have a HOLD_SIMVAR parameter, only toggle simvar, which will probably not result in the intended behavior. Try adding `<LEFT_LEAVE_CODE>0 (&gt;L:A32NX_OVHD_HYD_RAT_MAN_ON_IS_PRESSED)</LEFT_LEAVE_CODE>`, although this is not directly implemented by our template, the `ASOBO_GT_Push_Button_Airliner` template, that our generic template uses, has that feature.

[the remark this answers is not in the document]
Comment (crocket63) [reply flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/model/A320_NEO_INTERIOR.xml:1889]: It works great, thanks for the help! Elec button is corrected too then

File: src/systems/a320_systems/src/hydraulic.rs [hunk 27]
```diff
@@ -3802,5 +4537,53 @@ mod tests {
...
+                        } else {
+                            Potential::none()
+                        },
+                    );
+                }
+                _ => panic!("Bus not registered"),
+            }
+
+            supplied_power
+        }
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/hydraulic.rs:4587]: Does the potential origin or voltage really matter for testing? If not, then you can reduce this whole block of code to:

```rust
supplied_power.add(
    bus_id,
    if is_powered {
        Potential::single(
            PotentialOrigin::EngineGenerator(1),
            ElectricPotential::new::<volt>(115.),
         )
     } else {
         Potential::none()
     },
);
```

File: src/systems/a320_systems/src/lib.rs [hunk 28]
```diff
@@ -115,6 +115,8 @@ impl Aircraft for A320 {
             &self.hydraulic_overhead,
             &self.engine_fire_overhead,
             &self.landing_gear,
+            &self.emergency_electrical_overhead,
+            self.electrical.in_emergency_elec(),
```
Comment (davidwalschots) [inline src/systems/a320_systems/src/lib.rs:119]: **thought**: We might consider adding a trait for this too, such that we don't pass a bool around but something more meaningful. What do you think?

File: src/systems/a320_systems/src/hydraulic.rs [hunk 29]
```diff
@@ -295,10 +347,23 @@ impl A320Hydraulic {
...
         // Updating rat stowed pos on all frames in case it's used for graphics
         self.ram_air_turbine.update_position(&context.delta());
 
+        // Uses external conditions and momentary button: better to check each frame
+        self.ram_air_turbine_controller.update(
+            &overhead_panel,
+            rat_and_emer_gen_man_on.rat_and_emer_gen_man_on(),
+            is_in_emergency_elec,
+        );
+
```
[the remark this answers is not in the document]
Comment (crocket63) [reply src/systems/a320_systems/src/hydraulic.rs:366]: Makes definitly sense! Got lost rereading my code :p

File: src/systems/a320_systems/src/hydraulic.rs [hunk 30]
```diff
@@ -550,8 +616,12 @@ struct A320EngineDrivenPumpController {
 impl A320EngineDrivenPumpController {
     const MIN_ENGINE_OIL_PRESS_THRESHOLD_TO_INHIBIT_FAULT: f64 = 18.;
 
-    fn new(engine_number: usize) -> Self {
+    fn new(engine_number: usize, power_buses: Vec<ElectricalBusType>) -> Self {
+        assert!(!power_buses.is_empty() && power_buses.len() <= 2);
+
         Self {
+            is_powered: false,
+            powered_by: power_buses,
```
[the remark this answers is not in the document]
Comment (crocket63) [reply src/systems/a320_systems/src/hydraulic.rs:624]: Yes no use of this as implementation is dynamic now. I'll get back to fixes array if we struggle A LOT on performances...

File: src/systems/a320_systems/src/hydraulic.rs [hunk 31]
```diff
@@ -795,14 +892,24 @@ impl SimulationElement for A320YellowElectricPumpController {
...
+        self.is_powered = supplied_power.potential_of(&self.powered_by).is_powered()
+            || self
+                .should_activate_yellow_pump_for_cargo_door_operation
+                .output();
+    }
 }
 impl Default for A320YellowElectricPumpController {
     fn default() -> Self {
-        Self::new()
+        Self::new(ElectricalBusType::DirectCurrent(2))
```
[the remark this answers is not in the document]
Comment (crocket63) [reply src/systems/a320_systems/src/hydraulic.rs:906]: Damn clippy :D

File: src/systems/a320_systems/src/hydraulic.rs (hunk 27 above)
Comment (crocket63) [reply src/systems/a320_systems/src/hydraulic.rs:4587]: Nope indeed, was afraid to be slapped by engineers doing this, or to trigger elec issues with stupid values. If it's not the case simpler is better!

File: src/systems/a320_systems/src/hydraulic.rs [hunk 32]
```diff
@@ -795,14 +892,24 @@ impl SimulationElement for A320YellowElectricPumpController {
     fn write(&self, writer: &mut SimulatorWriter) {
         writer.write_bool("HYD_YELLOW_EPUMP_LOW_PRESS", self.is_pressure_low);
     }
+
+    fn receive_power(&mut self, supplied_power: &SuppliedPower) {
+        // Control of the pump is powered by dedicated bus OR manual operation of cargo door
+        self.is_powered = supplied_power.potential_of(&self.powered_by).is_powered()
+            || self
+                .should_activate_yellow_pump_for_cargo_door_operation
+                .output();
```
[the remark this answers is not in the document]
Comment (crocket63) [reply src/systems/a320_systems/src/hydraulic.rs:893]: I'd say it's a placeholder for some other relay powered from cargo door bay. @komp1821 is needed for elec details :D

File: src/systems/a320_systems/src/lib.rs (hunk 28 above)
Comment (crocket63) [reply src/systems/a320_systems/src/lib.rs]: Yup, done... well I think if my computer let me save a file :D

**Comment by crocket63** [comment]
Should be good! :D

**Review by davidwalschots — APPROVED** [review]
Go go go! 😀

**Comment by Kimbyeoungjang** [comment]
Quality Assurance Tester/Trainee Report

Discord: itkbj (김병장#7165)
Object of testing: #4810 
Tier of Testing: Tier 2
Date: 25/05/2021

Testing Process: Test all situation on reference

1. test Yellow ELEC Pump with connect GPU & with (DC BUS 2 or Cargo Door)
2. test Blue ELEC Pump with 1 engine start(for start BLUE ELEC PUMP automatically) & with DC ESS BUS
3. test RAT with HYD/RAT MAN ON pushbutton
4. test RAT with auto deploy (if speed >100kts and AC BUS 1+2 lost)
5. test RAT with EMER ELEC PWR/MAN ON pushbutton

Negatives: 
found bug with RAT deploy from HYD/RAT MAN ON pushbutton
it should push more than 1 sec to deply fully

Testing Results: Not passed.

Conclusions: 

Media: 
1. work as expect when connect GPU
image

2. work as expect when start 1 engine 
image

3-1. HYD/RAT MAN ON pushbutton on ground - work as expect. because we're on ground, RAT is not make any HYD
image

3-2. HYD/RAT MAN ON while on air.
note - issue on here. you should press RAT button more than 1 sec to deply RAT fully(look at that RAT _STOW_POSITION)

image

4. Lost AC 1 + 2 with speed > 100. work as expect
image

5. Press EMER ELEC PWR/MAN ON pushbutton. work as expect
note - issue also here. you should press EMER ELEC PWR/MAN ON pushbutton more than 1 sec to deply RAT fully

image


plus. I'll test right now after fix RAT bug.

**Comment by Kimbyeoungjang** [comment]
Non QA Associated Testing Report

Discord: itkbj (김병장#7165)
Object of testing: #4810
Tier of Testing: Tier 2
Date: 25/05/2021

Testing Process: 
test RAT things(others test finished)

perform full flight as flight plan

departure airport : RKSI
destination airport : RKPC
Cruise altitude : FL180

flight plan :
Direct MONSI - BULTI - MEKIL - GONAX - BEDES - ELPOS - MANGI - DALSU - DOTOL - KIDOS

arrival:
ILS 25-Y

Negatives: N/A

Testing Results: Passed.
Seems there is no performance impact. Thank you for your working! what a great function!

Media:
RAT bug fixed! - see the RAT_STOW_Position

image

**Comment by komp1821** [comment]
Looking very good overall. The pumps are correctly controlled and energized by their associated power supplies. RAT extension and pressurization (looked at the RAT RPM simvar and the blue system pressure) works. EDP solenoid supply is correct.

Found two minor issues.

### Issue 1: Blue pump activation logic
_What I did:_
1. External power on, normal electrical config at the gate.
2. Turn on BLUE PUMP OVRD on maintenance panel
3. Turn off BLUE PUMP on HYD(!!!) panel
4. Turn on BLUE PUMP on HYD panel.

_What I expected - (what happened) - conclusion:_
1. Power supplied - (Power supplied) - ok!
2. Blue pump turns on - (Blue pump turned on) - ok!
3. Blue pump turns off - (Blue pump turned off) - ok!
4. Blue pump remains off - (Blue pump turns on again) - **not ok!**

The OVRD pushbutton is a momentary pushbutton that latches a relay when pushed. The relay should open again when the BLUE PUMP pushbutton on the HYD panel is released. The problem: In our sim, the OVRD pushbutton is a latching pushbutton. I think this needs to be changed before we can proceed with this. Evidence/references, if needed, via Discord.

### Issue 2: Brake accu gauge behavior with power off
When pressurizing the yellow pump from cargo door activation, the brake accu is filled. Confirmed by looking at `A32NX_HYD_BRAKE_ALTN_ACC_PRESS` That is correct and intended. The brake accu gauge however does not show this.
This seems to be a more general issue. When all power is off, the brake accu gauge needles go to 0. They shouldn't.

**Comment by crocket63** [comment]
> Looking very good overall. The pumps are correctly controlled and energized by their associated power supplies. RAT extension and pressurization (looked at the RAT RPM simvar and the blue system pressure) works. EDP solenoid supply is correct.
> 
> Found two minor issues.
> 
> ### Issue 1: Blue pump activation logic
> _What I did:_
> 
> 1. External power on, normal electrical config at the gate.
> 2. Turn on BLUE PUMP OVRD on maintenance panel
> 3. Turn off BLUE PUMP on HYD(!!!) panel
> 4. Turn on BLUE PUMP on HYD panel.
> 
> _What I expected - (what happened) - conclusion:_
> 
> 1. Power supplied - (Power supplied) - ok!
> 2. Blue pump turns on - (Blue pump turned on) - ok!
> 3. Blue pump turns off - (Blue pump turned off) - ok!
> 4. Blue pump remains off - (Blue pump turns on again) - **not ok!**
> 
> The OVRD pushbutton is a momentary pushbutton that latches a relay when pushed. The relay should open again when the BLUE PUMP pushbutton on the HYD panel is released. The problem: In our sim, the OVRD pushbutton is a latching pushbutton. I think this needs to be changed before we can proceed with this. Evidence/references, if needed, via Discord.
> 
> ### Issue 2: Brake accu gauge behavior with power off
> When pressurizing the yellow pump from cargo door activation, the brake accu is filled. Confirmed by looking at `A32NX_HYD_BRAKE_ALTN_ACC_PRESS` That is correct and intended. The brake accu gauge however does not show this.
> This seems to be a more general issue. When all power is off, the brake accu gauge needles go to 0. They shouldn't.

Thanks a lot!
I propose we try to fix the blue override in a separate PR, and look into the brake gauge power issue right now.
